{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2411d2a7",
   "metadata": {},
   "source": [
    "# Week 3 Python Code Assignment\n",
    "BIOL 51000 Data Systems in the Life Sciences\n",
    "\n",
    "November 2022\n",
    "\n",
    "Victoria Griffin"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec593279",
   "metadata": {},
   "source": [
    "Python code for measuring conservation with generated random matrix that contains seven rows of 21 column vectors with A, C, T and G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4797e6e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AGTACGTATCGATGGACCTTA', 'ATCCGTATGCCCTTAATTAGA', 'ATGCGTGCAATGCTATTCCTA', 'TTGCTATTAATGCCCTCCAAA', 'ATCGTAGCTAGACTTTAGTAA', 'ATCGTAGTCTGATTTTAACCA', 'ATCGTAGTCTGATCCCAATTA']\n"
     ]
    }
   ],
   "source": [
    "# Random matrix\n",
    "\n",
    "randMat = ['AGTACGTATCGATGGACCTTA',\n",
    "           'ATCCGTATGCCCTTAATTAGA',\n",
    "           'ATGCGTGCAATGCTATTCCTA',\n",
    "           'TTGCTATTAATGCCCTCCAAA',\n",
    "           'ATCGTAGCTAGACTTTAGTAA',\n",
    "           'ATCGTAGTCTGATTTTAACCA', \n",
    "           'ATCGTAGTCTGATCCCAATTA']\n",
    "\n",
    "print(randMat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "642e5bb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'G': {'G': 8, 'C': -5, 'A': -5, 'T': -5, 'N': 0}, 'C': {'G': -5, 'C': 8, 'A': -5, 'T': -5, 'N': 0}, 'A': {'G': -5, 'C': -5, 'A': 8, 'T': -5, 'N': 0}, 'T': {'G': -5, 'C': -5, 'A': -5, 'T': 8, 'N': 0}, 'N': {'G': 0, 'C': 0, 'A': 0, 'T': 0, 'N': 0}}\n"
     ]
    }
   ],
   "source": [
    "#Reference matrix used for comparing bases on two sequences â†’ match is 8, mismatch is -5, no sequence data is 0\n",
    "\n",
    "DNA = {'G': { 'G': 8, 'C':-5, 'A':-5, 'T':-5, 'N':0 },\n",
    "       'C': { 'G':-5, 'C': 8, 'A':-5, 'T':-5, 'N':0 },\n",
    "       'A': { 'G':-5, 'C':-5, 'A': 8, 'T':-5, 'N':0 },\n",
    "       'T': { 'G':-5, 'C':-5, 'A':-5, 'T': 8, 'N':0 },\n",
    "       'N': { 'G': 0, 'C': 0, 'A': 0, 'T': 0, 'N':0 }} \n",
    "\n",
    "print(DNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "539e422f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "I work!\n"
     ]
    }
   ],
   "source": [
    "# In order to calculate similarity scores, we must first define the profile function from Chapter 13\n",
    "\n",
    "def profile(alignment):\n",
    "  \n",
    "  n = len(alignment[0]) # Defines the length of the alignment\n",
    "  numSequences = float(len(alignment)) # Finds the number of sequences\n",
    "  alignInfo = [] # Creates a list for storing alignment information\n",
    "  \n",
    "  for i in range(n):\n",
    "    counts = {} # Counts directory to store the counts for nucleotide storage\n",
    "    \n",
    "    # Loops through each sequence to determine frequency at every position\n",
    "    for seq in alignment:\n",
    "      letter = seq[i]\n",
    "      if letter == '-':\n",
    "        continue\n",
    "    \n",
    "      counts[letter] = counts.get(letter, 0) + 1\n",
    "    \n",
    "    # Determines the num of times a nucleotide falls in a specific position by /total count by num of sequences\n",
    "    for letter in counts:\n",
    "      counts[letter] /= nSeq\n",
    "    \n",
    "    # Stores calculated positional data\n",
    "    alignInfo.append(counts)\n",
    "\n",
    "  return alignInfo\n",
    "\n",
    "print (\"I work!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c4b597f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "I love Biology!\n"
     ]
    }
   ],
   "source": [
    "# Creating a function that will calculate similarity scores for the sequences we listed above in randMat\n",
    "\n",
    "def findConservation (align, matrix):\n",
    "    conservation = [] # blank list for holding scores\n",
    "    profileAlign = profile(alignment)\n",
    "    \n",
    "    # Loop through all dicts of the profile\n",
    "    for compositionDict in profileAlign:\n",
    "        \n",
    "        items  = list(compositionDict.items()) # Need .items() when using Python 3\n",
    "        \n",
    "        # Sort items list by composition value, so we can get a rankings\n",
    "        items.sort (key=lambda x: x[1])\n",
    "        \n",
    "        score = 0.0\n",
    "        \n",
    "        # Loop through items that now contain residue codes and compsition values\n",
    "        # Score will be * of compositional fractions and the similarity score from DNA matrix\n",
    "        for residueA, compositionA in items:\n",
    "            for residueB, compositionB in items:\n",
    "                score += compositionA * compositionB * simMatrix[residueA][residueB]\n",
    "        \n",
    "        # Will give us the best residue code from the end of our sorted list and use it as both keys in matrix dict\n",
    "        bestAlginLetter = items [-1][0]\n",
    "        maxScore = matrix[bestAlignLetter][bestAlignLetter]\n",
    "        \n",
    "        # Score is scaled by dividing it by the maximum value and adding it to conservation list\n",
    "        score /= maxScore\n",
    "        \n",
    "        conservation.append(score)\n",
    "    \n",
    "    return conservation\n",
    "\n",
    "print(\"I love Biology!\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "22226db2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculated Similarity Scores:\n",
      "[0.6020408163265305, 0.6020408163265305, 0.07142857142857145, 0.005102040816326536, 0.07142857142857145, 0.07142857142857145, 0.07142857142857145, 0.07142857142857145, -0.19387755102040816, -0.061224489795918324, 0.07142857142857145, 0.07142857142857145, 0.20408163265306123, 0.07142857142857145, -0.19387755102040816, 0.07142857142857145, -0.061224489795918324, -0.12755102040816324, -0.061224489795918324, -0.12755102040816324, 1.0]\n"
     ]
    }
   ],
   "source": [
    "# Putting it all together\n",
    "\n",
    "\n",
    "randMat = ['AGTACGTATCGATGGACCTTA',\n",
    "           'ATCCGTATGCCCTTAATTAGA',\n",
    "           'ATGCGTGCAATGCTATTCCTA',\n",
    "           'TTGCTATTAATGCCCTCCAAA',\n",
    "           'ATCGTAGCTAGACTTTAGTAA',\n",
    "           'ATCGTAGTCTGATTTTAACCA', \n",
    "           'ATCGTAGTCTGATCCCAATTA']\n",
    "\n",
    "\n",
    "DNA = {'G': { 'G': 8, 'C':-5, 'A':-5, 'T':-5, 'N':0 },\n",
    "       'C': { 'G':-5, 'C': 8, 'A':-5, 'T':-5, 'N':0 },\n",
    "       'A': { 'G':-5, 'C':-5, 'A': 8, 'T':-5, 'N':0 },\n",
    "       'T': { 'G':-5, 'C':-5, 'A':-5, 'T': 8, 'N':0 },\n",
    "       'N': { 'G': 0, 'C': 0, 'A': 0, 'T': 0, 'N':0 }} \n",
    "\n",
    "\n",
    "def profile(alignment):\n",
    "  \n",
    "  n = len(alignment[0]) # Defines the length of the alignment\n",
    "  numSequences = float(len(alignment)) # Finds the number of sequences\n",
    "  alignInfo = [] # Creates a list for storing alignment information\n",
    "  \n",
    "  for i in range(n):\n",
    "    counts = {} # Counts directory to store the counts for nucleotide storage\n",
    "    \n",
    "    # Loops through each sequence to determine frequency at every position\n",
    "    for seq in alignment:\n",
    "      letter = seq[i]\n",
    "      if letter == '-':\n",
    "        continue\n",
    "    \n",
    "      counts[letter] = counts.get(letter, 0) + 1\n",
    "    \n",
    "    # Determines the num of times a nucleotide falls in a specific position by /total count by num of sequences\n",
    "    for letter in counts:\n",
    "      counts[letter] /= numSequences\n",
    "    \n",
    "    # Stores calculated positional data\n",
    "    alignInfo.append(counts)\n",
    "\n",
    "  return alignInfo\n",
    "\n",
    "# Creating a function that will calculate similarity scores for the sequences we listed above in randMat\n",
    "\n",
    "def findConservation (align, matrix):\n",
    "    conservation = [] # blank list for holding scores\n",
    "    profileAlign = profile(align)\n",
    "    \n",
    "    # Loop through all dicts of the profile\n",
    "    for compositionDict in profileAlign:\n",
    "        \n",
    "        items  = list(compositionDict.items()) # Need .items() when using Python 3\n",
    "        \n",
    "        # Sort items list by composition value, so we can get a rankings\n",
    "        items.sort (key=lambda x: x[1])\n",
    "        \n",
    "        score = 0.0\n",
    "        \n",
    "        # Loop through items that now contain residue codes and compsition values\n",
    "        # Score will be * of compositional fractions and the similarity score from DNA matrix\n",
    "        for residueA, compositionA in items:\n",
    "            for residueB, compositionB in items:\n",
    "                score += compositionA * compositionB * matrix[residueA][residueB]\n",
    "        \n",
    "        # Will give us the best residue code from the end of our sorted list and use it as both keys in matrix dict\n",
    "        bestAlignLetter = items [-1][0]\n",
    "        maxScore = matrix[bestAlignLetter][bestAlignLetter]\n",
    "        \n",
    "        # Score is scaled by dividing it by the maximum value and adding it to conservation list\n",
    "        score /= maxScore\n",
    "        \n",
    "        conservation.append(score)\n",
    "    \n",
    "    return conservation\n",
    "\n",
    "print('Calculated Similarity Scores:')\n",
    "print(findConservation(randMat, DNA))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc77f6b7",
   "metadata": {},
   "source": [
    "Python code for calculating substitution matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b962edbd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'G': {'G': 10, 'C': 0, 'A': -6, 'T': -2, 'N': 0}, 'C': {'G': 0, 'C': 5, 'A': -4, 'T': 0, 'N': 0}, 'A': {'G': -6, 'C': -4, 'A': 8, 'T': -5, 'N': 0}, 'T': {'G': -2, 'C': 0, 'A': -5, 'T': 4, 'N': 0}, 'N': {'G': 0, 'C': 0, 'A': 0, 'T': 0, 'N': 0}}\n",
      "\n",
      "COOL!\n"
     ]
    }
   ],
   "source": [
    "# Need to import log function first from math\n",
    "from math import log\n",
    "\n",
    "# define function calculateSubMatrix for calculating substitution matrices\n",
    "def calculateSubMatrix(alignment, alphabet, maxValue, smoothData=5):\n",
    "\n",
    "  subMatrix = {} # This empty dict will contain substitution matrix\n",
    "  counts = {} # Dict containing residue counts\n",
    "  \n",
    "  # Will measure number of residues and number of substitutions of one residue to another\n",
    "  for letterA in alphabet:\n",
    "    subDict = {}\n",
    "    \n",
    "    for letterB in alphabet:\n",
    "      subDict[letterB] = 0\n",
    "  \n",
    "    subMatrix[letterA] = subDict\n",
    "    counts[letterA] = 0\n",
    "  \n",
    "  # Will need these for calculations later on - 0.0 because division will occur\n",
    "  totalResidue = 0.0\n",
    "  totalSub = 0.0\n",
    "  \n",
    "  # Most important part of the function:\n",
    "  # Loops through all input alignments and collects substitutional data\n",
    "  for align in alignment:\n",
    " \n",
    "    numberPosition = len(align[0]) #saves the length of the sequence in variable - 0 is first sequence\n",
    "\n",
    "    for i in range(numberPosition):\n",
    "      \n",
    "      # Initialize list for collecting residues \n",
    "      letters = []\n",
    "      \n",
    "      for sequence in align:\n",
    "        \n",
    "        # If gap, skip and move on\n",
    "        letter = sequence[i]\n",
    "        if letter == '-':\n",
    "          continue\n",
    "        \n",
    "        # Add resdiue to list\n",
    "        letters.append(letter)\n",
    "\n",
    "      # Once residues are collected, go throught again to get possible pairs \n",
    "      for letterA in letters:\n",
    "        counts[letterA] += 1 # Count for each letter increases by one\n",
    "      \n",
    "        for letterB in letters:          \n",
    "          subMatrix[letterA][letterB] += 1 # Add to matrix\n",
    "\n",
    "      numLetters = len(letters) # length so far, not including gaps\n",
    "      totalResidue += numLetters # Add length of letters to totalResidue   \n",
    "      totalSub += numLetters * numLetters # Substitution total increased by numLetters^2\n",
    "\n",
    "  # Blank dict for saving average composition calculations \n",
    "  avgComposition = {}    \n",
    "  \n",
    "  # Calculate average composition   \n",
    "  for letter in alphabet:\n",
    "    avgComposition[letter] = counts[letter]/totalResidue      \n",
    "\n",
    "  # Initialize variable maximumScore  \n",
    "  maximumScore = None\n",
    "  for residueA in alphabet:\n",
    "    for residueB in alphabet:\n",
    "\n",
    "      expected = avgComposition[residueA] * avgComposition[residueB]\n",
    "      \n",
    "      if not expected: # If calculation is 0, skip\n",
    "        continue\n",
    "      \n",
    "      # pair count of combination \n",
    "      observed = subMatrix[residueA][residueB]\n",
    "      weight = 1.0 / (1.0+(observed/smoothData)) \n",
    "\n",
    "      # Converted into probability\n",
    "      observed /= totalSub\n",
    "      observed = weight*expected + (1-weight)*observed #Apply weight calculated from above\n",
    "\n",
    "      # log-odds scores is calculated by taking log of probability ratio \n",
    "      logOdds = log(observed/expected)\n",
    "      \n",
    "      # If a new maxScore is found, we will add score to the matrix\n",
    "      if (maximumScore is None) or (logOdds>maximumScore):\n",
    "        maximumScore = logOdds\n",
    "      \n",
    "      subMatrix[residueA][residueB] = logOdds\n",
    "  \n",
    "  # find absolute value of maxscore\n",
    "  maxScore = abs(maximumScore)\n",
    " \n",
    "  #Scale the log-odds values in substitution matrix and convert them to integers\n",
    "  for resA in alphabet:\n",
    "    for resB in alphabet:\n",
    "      subMatrix[resA][resB] = int(maxValue*subMatrix[resA][resB]/maximumScore)\n",
    "\n",
    "  return subMatrix\n",
    "\n",
    "\n",
    "#Test function\n",
    "\n",
    "randMat = ['AGTACGTATCGATGGACCTTA',\n",
    "           'ATCCGTATGCCCTTAATTAGA',\n",
    "           'ATGCGTGCAATGCTATTCCTA',\n",
    "           'TTGCTATTAATGCCCTCCAAA',\n",
    "           'ATCGTAGCTAGACTTTAGTAA',\n",
    "           'ATCGTAGTCTGATTTTAACCA', \n",
    "           'ATCGTAGTCTGATCCCAATTA']\n",
    "\n",
    "\n",
    "DNA = {'G': { 'G': 8, 'C':-5, 'A':-5, 'T':-5, 'N':0 },\n",
    "       'C': { 'G':-5, 'C': 8, 'A':-5, 'T':-5, 'N':0 },\n",
    "       'A': { 'G':-5, 'C':-5, 'A': 8, 'T':-5, 'N':0 },\n",
    "       'T': { 'G':-5, 'C':-5, 'A':-5, 'T': 8, 'N':0 },\n",
    "       'N': { 'G': 0, 'C': 0, 'A': 0, 'T': 0, 'N':0 }} \n",
    "\n",
    "nucleotides = DNA.keys()\n",
    "\n",
    "print(calculateSubMatrix([randMat,], nucleotides, 10))\n",
    "print('')\n",
    "print('COOL!')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6e9e61a",
   "metadata": {},
   "source": [
    "Python code for calculating distance matrices (refers to Phylogenetic Trees)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d06debb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence Alignment\n"
     ]
    }
   ],
   "source": [
    "# FIRST NEED TO USE SEQUENCE ALIGNMENT FUNCTION FROM WEEK 1 ASSIGNMENT\n",
    "\n",
    "# Define function called seqAlign that performs a simple pairwise alignment\n",
    "def seqAlign(seqA, seqB, simMatrix=DNA, insert=3, extend=4): \n",
    "    # gap is penalty of 3, and extension of gap is a penalty of 4\n",
    "\n",
    "  \"\"\"Variables numI and numJ represent the length of the rows and columns of the comparison grid â†’ \n",
    "     needs to be one length longer than each sequence\"\"\"\n",
    "  numI = len(seqA) + 1\n",
    "  numJ = len(seqB) + 1\n",
    "\n",
    "  # Create two matrices. One for holding the scores and one for holding the route taken\n",
    "  scoreMatrix = [[0] * numJ for x in range(numI)]\n",
    "  routeMatrix = [[0] * numJ for x in range(numI)]\n",
    "  \n",
    "  \"\"\"I AM NOT SURE IF THIS NEXT SECTION IS CORRECT --> I AM TRYING TO APPLY 0, -3, -9, etc. TO THE OUTER \n",
    "     ROW AND COLUMN OF THE MATRIX LIKE WE DID DURING THE LECTURE.\"\"\"\n",
    "\n",
    "   # Variables needed to label matrix with correct outer column and row (eg. 0, -3, -6, -9, -12â€¦)\n",
    "  leftColumn = 0\n",
    "  topRow = 0\n",
    "\n",
    "  # Make leftmost column have score start with 0 and decrease by 3 for each iteration\n",
    "  for i in range(1, numI):\n",
    "    routeMatrix[i][0] = leftColumn\n",
    "    leftColumn -= 3\n",
    "    \n",
    "  # Make topmost row have score start with 0 and decrease by 3 for each iteration\n",
    "  for j in range(1, numJ):\n",
    "    routeMatrix[0][j] = topRow\n",
    "    topRow -= 3\n",
    "  \n",
    " # Nested loop used to fill in rest of the matrix \n",
    "  for i in range(1, numI):\n",
    "    for j in range(1, numJ):\n",
    "      \n",
    "      # Penalty of a gap is 3 like our example from class \n",
    "      penalty1 = insert\n",
    "      penalty2 = insert\n",
    "      \n",
    "      #NOT SURE HOW TO FIX THIS THOUGH TO MATCH THE 0,-3, -6â€¦.THIS IS WHAT I CAME UP WITH\n",
    "      # Check to see if previous is a gap\n",
    "      if routeMatrix[i-1][j] == routeMatrix[0][j]:\n",
    "        penalty1 = extend\n",
    "        \n",
    "      elif routeMatrix[i][j-1] == routeMatrix[i][0]:\n",
    "        penalty2 = extend\n",
    "       \n",
    "       # Look up similarity score of bases for seqA and seqB \n",
    "      similarity = simMatrix[ seqA[i-1] ][ seqB[j-1] ]\n",
    "      \n",
    "      # Make a list of three possible scores (left-up diagonal, up, and left)\n",
    "      paths = [scoreMatrix[i-1][j-1] + similarity, # Route 0\n",
    "                   scoreMatrix[i-1][j] - penalty1, # Route 1\n",
    "                   scoreMatrix[i][j-1] - penalty2] # Route 2                     \n",
    "      \n",
    "      # The best of the three scores are assigned in â€˜bestâ€™ and the index of best score is assigned to route\n",
    "      best = max(paths)\n",
    "      route = paths.index(best)           \n",
    "\n",
    "      # The scores and indexes for each iteration are saved in the matrix \n",
    "      scoreMatrix[i][j] = best\n",
    "      routeMatrix[i][j] = route\n",
    "\n",
    "#___________________________________________Working Backwards______________________________________________#\n",
    "   \n",
    "   #  Initialize blank lists to hold the optimal sequence alignments for seqA and seqB    \n",
    "  alignA = []\n",
    "  alignB = []\n",
    "\n",
    "  # Assigning variables to implement into while loop below for working backwards\n",
    "  i = numI-1\n",
    "  j = numJ-1\n",
    "  score = scoreMatrix[i][j]\n",
    "\n",
    "\n",
    "  while i > 0 or j > 0:\n",
    "    route = routeMatrix[i][j]    \n",
    "\n",
    "     # Determining which of the three routes had the highest score\n",
    "    if route == 0: # Diagonal\n",
    "      alignA.append( seqA[i-1] )\n",
    "      alignB.append( seqB[j-1] )\n",
    "      i -= 1\n",
    "      j -= 1\n",
    "\n",
    "    elif route == 1: # Gap in seqB\n",
    "      alignA.append( seqA[i-1] )\n",
    "      alignB.append( '-' )\n",
    "      i -= 1      \n",
    "\n",
    "    elif route == 2: # Gap in seqA\n",
    "      alignA.append( '-' )\n",
    "      alignB.append( seqB[j-1] ) \n",
    "      j -= 1\n",
    "  \n",
    "   # Reverse the alignment lists since we originally stored them backwards, now we want them forwards \n",
    "  alignA.reverse()\n",
    "  alignB.reverse()\n",
    "  alignA = ''.join(alignA)\n",
    "  alignB = ''.join(alignB)\n",
    "\n",
    "  return score, alignA, alignB \n",
    "\n",
    "print ('Sequence Alignment')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d085005b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculate Sequence Similarity\n"
     ]
    }
   ],
   "source": [
    "# THEN NEED TO USE CALCULATE SEQUENCE SIMILARITY FUNCTION\n",
    "\n",
    "# TOOK FROM BOOK\n",
    "def calcSeqSimilarity(seqA, seqB, simMatrix):\n",
    "\n",
    "  numPlaces = min(len(seqA), len(seqB))\n",
    "  \n",
    "  totalScore = 0.0\n",
    "  \n",
    "  for i in range(numPlaces):\n",
    "    \n",
    "    residueA = seqA[i]\n",
    "    residueB = seqB[i]\n",
    "  \n",
    "    totalScore += simMatrix[residueA][residueB]\n",
    "\n",
    "  return totalScore\n",
    "\n",
    "print('Calculate Sequence Similarity')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d96b1fc9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Distance Matrix\n"
     ]
    }
   ],
   "source": [
    "# NOW WE CAN CALCUATE THE DISTANCE MATRICES\n",
    "\n",
    "#First define a function that will create a simple distance matrix\n",
    "def getDistanceMatrix(seqs, simMatrix):\n",
    "\n",
    "  numbers = len(seqs) #variable numbers for storing sequence lengths\n",
    "  matrix = [[0.0] * numbers for x in range(numbers)]\n",
    "  maxScores = [calcSeqSimilarity(x, x, simMatrix) for x in seqs]\n",
    "  \n",
    "  # Sequences are aligned using an input subsititution matrix\n",
    "  for i in range(numbers-1):\n",
    "    seqA = seqs[i]\n",
    "  \n",
    "    for j in range(i+1,numbers):\n",
    "      seqB = seqs[j]\n",
    "      \n",
    "      # Alignment is passed back with a score which then gets converted into a distance\n",
    "      score, alignA, alignB = seqAlign(seqA, seqB, simMatrix)\n",
    "      maxScore = max(maxScores[i],maxScores[j])\n",
    "      dist = maxScore - score\n",
    "      \n",
    "      \n",
    "      matrix[i][j] = dist\n",
    "      matrix[j][i] = dist\n",
    "\n",
    "  return matrix\n",
    "\n",
    "print('Distance Matrix')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bd5c13bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Distance Matrix\n",
      "['0.00', '133.00', '120.00', '134.00', '119.00', '114.00', '101.00']\n",
      "['133.00', '0.00', '123.00', '118.00', '103.00', '124.00', '100.00']\n",
      "['120.00', '123.00', '0.00', '126.00', '111.00', '114.00', '114.00']\n",
      "['134.00', '118.00', '126.00', '0.00', '142.00', '138.00', '125.00']\n",
      "['119.00', '103.00', '111.00', '142.00', '0.00', '76.00', '100.00']\n",
      "['114.00', '124.00', '114.00', '138.00', '76.00', '0.00', '65.00']\n",
      "['101.00', '100.00', '114.00', '125.00', '100.00', '65.00', '0.00']\n"
     ]
    }
   ],
   "source": [
    "# TEST DISTANCE MATRIX\n",
    "\n",
    "seqData = ['AGTACGTATCGATGGACCTTA',\n",
    "           'ATCCGTATGCCCTTAATTAGA',\n",
    "           'ATGCGTGCAATGCTATTCCTA',\n",
    "           'TTGCTATTAATGCCCTCCAAA',\n",
    "           'ATCGTAGCTAGACTTTAGTAA',\n",
    "           'ATCGTAGTCTGATTTTAACCA', \n",
    "           'ATCGTAGTCTGATCCCAATTA']\n",
    "\n",
    "DNA = {'G': { 'G': 8, 'C':-5, 'A':-5, 'T':-5, 'N':0 },\n",
    "       'C': { 'G':-5, 'C': 8, 'A':-5, 'T':-5, 'N':0 },\n",
    "       'A': { 'G':-5, 'C':-5, 'A': 8, 'T':-5, 'N':0 },\n",
    "       'T': { 'G':-5, 'C':-5, 'A':-5, 'T': 8, 'N':0 },\n",
    "       'N': { 'G': 0, 'C': 0, 'A': 0, 'T': 0, 'N':0 }} \n",
    "\n",
    "def seqAlign(seqA, seqB, simMatrix=DNA, insert=8, extend=4): \n",
    " \n",
    "\n",
    "  \"\"\"Variables numI and numJ represent the length of the rows and columns of the comparison grid â†’ \n",
    "     needs to be one length longer than each sequence\"\"\"\n",
    "  numI = len(seqA) + 1\n",
    "  numJ = len(seqB) + 1\n",
    "\n",
    "  # Create two matrices. One for holding the scores and one for holding the route taken\n",
    "  scoreMatrix = [[0] * numJ for x in range(numI)]\n",
    "  routeMatrix = [[0] * numJ for x in range(numI)]\n",
    "  \n",
    "  # Make leftmost column have score start with 0 and decrease by 3 for each iteration\n",
    "  for i in range(1, numI):\n",
    "    routeMatrix[i][0] = 1\n",
    "   \n",
    "    \n",
    "  # Make topmost row have score start with 0 and decrease by 3 for each iteration\n",
    "  for j in range(1, numJ):\n",
    "    routeMatrix[0][j] = 2\n",
    "  \n",
    " # Nested loop used to fill in rest of the matrix \n",
    "  for i in range(1, numI):\n",
    "    for j in range(1, numJ):\n",
    "      \n",
    "      # Penalty of a gap is 3 like our example from class \n",
    "      penalty1 = insert\n",
    "      penalty2 = insert\n",
    "      \n",
    "      #NOT SURE HOW TO FIX THIS THOUGH TO MATCH THE 0,-3, -6â€¦.THIS IS WHAT I CAME UP WITH\n",
    "      # Check to see if previous is a gap\n",
    "      if routeMatrix[i-1][j] == 1:\n",
    "        penalty1 = extend\n",
    "        \n",
    "      elif routeMatrix[i][j-1] == 2:\n",
    "        penalty2 = extend\n",
    "       \n",
    "       # Look up similarity score of bases for seqA and seqB \n",
    "      similarity = simMatrix[seqA[i-1] ][seqB[j-1] ]\n",
    "      \n",
    "      # Make a list of three possible scores (left-up diagonal, up, and left)\n",
    "      paths = [scoreMatrix[i-1][j-1] + similarity, # Route 0\n",
    "               scoreMatrix[i-1][j] - penalty1, # Route 1\n",
    "               scoreMatrix[i][j-1] - penalty2] # Route 2                     \n",
    "      \n",
    "      # The best of the three scores are assigned in â€˜bestâ€™ and the index of best score is assigned to route\n",
    "      best = max(paths)\n",
    "      route = paths.index(best)           \n",
    "\n",
    "      # The scores and indexes for each iteration are saved in the matrix \n",
    "      scoreMatrix[i][j] = best\n",
    "      routeMatrix[i][j] = route\n",
    "   \n",
    "   #  Initialize blank lists to hold the optimal sequence alignments for seqA and seqB    \n",
    "  alignA = []\n",
    "  alignB = []\n",
    "\n",
    "  # Assigning variables to implement into while loop below for working backwards\n",
    "  i = numI-1\n",
    "  j = numJ-1\n",
    "  score = scoreMatrix[i][j]\n",
    "\n",
    "  while i > 0 or j > 0:\n",
    "    route = routeMatrix[i][j]    \n",
    "\n",
    "     # Determining which of the three routes had the highest score\n",
    "    if route == 0: # Diagonal\n",
    "      alignA.append( seqA[i-1] )\n",
    "      alignB.append( seqB[j-1] )\n",
    "      i -= 1\n",
    "      j -= 1\n",
    "\n",
    "    elif route == 1: # Gap in seqB\n",
    "      alignA.append(seqA[i-1] )\n",
    "      alignB.append('-')\n",
    "      i -= 1      \n",
    "\n",
    "    elif route == 2: # Gap in seqA\n",
    "      alignA.append('-')\n",
    "      alignB.append(seqB[j-1] ) \n",
    "      j -= 1\n",
    "  \n",
    "   # Reverse the alignment lists since we originally stored them backwards, now we want them forwards \n",
    "  alignA.reverse()\n",
    "  alignB.reverse()\n",
    "  alignA = ''.join(alignA)\n",
    "  alignB = ''.join(alignB)\n",
    "\n",
    "  return score, alignA, alignB \n",
    "\n",
    "def calcSeqSimilarity(seqA, seqB, simMatrix):\n",
    "\n",
    "  numPlaces = min(len(seqA), len(seqB))\n",
    "  \n",
    "  totalScore = 0.0\n",
    "  \n",
    "  for i in range(numPlaces):\n",
    "    \n",
    "    residueA = seqA[i]\n",
    "    residueB = seqB[i]\n",
    "  \n",
    "    totalScore += simMatrix[residueA][residueB]\n",
    "\n",
    "  return totalScore\n",
    "\n",
    "def getDistanceMatrix(seqs, simMatrix):\n",
    "\n",
    "  numbers = len(seqs) #variable numbers for storing sequence lengths\n",
    "  matrix = [[0.0] * numbers for x in range(numbers)]\n",
    "  maxScores = [calcSeqSimilarity(x, x, simMatrix) for x in seqs]\n",
    "  \n",
    "  # Sequences are aligned using an input subsititution matrix\n",
    "  for i in range(numbers-1):\n",
    "    seqA = seqs[i]\n",
    "  \n",
    "    for j in range(i+1,numbers):\n",
    "      seqB = seqs[j]\n",
    "      \n",
    "      # Alignment is passed back with a score which then gets converted into a distance\n",
    "      score, alignA, alignB = seqAlign(seqA, seqB, simMatrix)\n",
    "      maxScore = max(maxScores[i],maxScores[j])\n",
    "      dist = maxScore - score\n",
    "      \n",
    "      matrix[i][j] = dist\n",
    "      matrix[j][i] = dist\n",
    "\n",
    "  return matrix\n",
    "\n",
    "distMatrix = getDistanceMatrix(seqData, DNA)\n",
    "\n",
    "print ('Distance Matrix')\n",
    "for row in distMatrix:\n",
    "    print(['%.2f' % x for x in row])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67694895",
   "metadata": {},
   "source": [
    "This entire next section inludes the other three functions needed to take all the raw data and put it into a completed tree. I am not sure if I needed to do that or not, but I did."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e3d24ecd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "I may have done too much work for this problem...not sure\n"
     ]
    }
   ],
   "source": [
    "# Next we define a function to find out which element of a distance matrix represents the closest sequences\n",
    "# This means finding which sequences are should be joined next when creating a tree\n",
    "def getJoinPair(distMatrix):\n",
    "\n",
    "  numbers = len(distMatrix) #variable for holding length of matrix\n",
    "  \n",
    "  minQ = None\n",
    "  joinPair = None\n",
    "\n",
    "  # nested loop for iterating through matrix  \n",
    "  for i in range(numbers-1):\n",
    "    sumRow = sum(distMatrix[i])\n",
    "  \n",
    "    for j in range(i+1, numbers):\n",
    "      sumCol = sum(distMatrix[j])\n",
    "    \n",
    "      dist = distMatrix[i][j]\n",
    "      q = (n-2)*dist - sumRow - sumCol # Most important here is how we calculate q\n",
    "      \n",
    "      if (minQ is None) or (q < minQ):\n",
    "        minQ = q\n",
    "        joinPair = [i,j] # If smallest so far, we save it as the joined pair aka closest sequences\n",
    "\n",
    "  return joinPair\n",
    "\n",
    "\n",
    "# Next function is to determine distance from a point on tree to the branch point...helps find distance to nodes\n",
    "def getDistToJunction(distMatrix, i, j):\n",
    "  \n",
    "  numbers = len(distMatrix) #variable for holding length of matrix\n",
    "  row = distMatrix[i] # i is row\n",
    "  column = distMatrix[j] # j is column\n",
    " \n",
    "  dist = distMatrix[i][j] + (sum(row)-sum(column))/(numbers-2) # calculate distance\n",
    "  dist *= 0.5 #take that and divide by 2\n",
    "\n",
    "  return dist\n",
    "\n",
    "# Last function builds the tree\n",
    "def neighbourJoinTree(distMatrix):\n",
    " \n",
    "  joinOrder = [] #empty list\n",
    "  numbers = len(distMatrix) #variable for holding length of matrix\n",
    "  tree = list(range(n))  # list needed for Python 3\n",
    "  \n",
    "  # Using a while loop to iterate while numbers is greater than 2\n",
    "  while numbers > 2:\n",
    "\n",
    "    x, y = getJoinPair(distMatrix)\n",
    "\n",
    "    # New branch point is joined\n",
    "    node = (tree[x], tree[y])\n",
    "    joinOrder.append(node)\n",
    "    tree.append(node)\n",
    "\n",
    "    del tree[y]\n",
    "    del tree[x]\n",
    "\n",
    "    distX = getDistToJunction(distMatrix, x, y)\n",
    "    distY = getDistToJunction(distMatrix, y, x)\n",
    "  \n",
    "    distMatrix.append([0] * (numbers+1))\n",
    "\n",
    "    for i in range(numbers):\n",
    "      if i not in (x,y):\n",
    "\n",
    "        dist = (distMatrix[x][i]-distX) + (distMatrix[y][i]-distY)\n",
    "        dist *= 0.5\n",
    "  \n",
    "        distMatrix[i].append(dist)\n",
    "        distMatrix[nnumbers][i] = dist\n",
    "\n",
    "    del distMatrix[y]\n",
    "    del distMatrix[x]\n",
    "  \n",
    "    for row in distMatrix:\n",
    "      del row[y]\n",
    "      del row[x]\n",
    "\n",
    "    numbers -= 1\n",
    "\n",
    "  tree = tuple(tree) #Change tree to a tuple\n",
    "  joinOrder.append(tree)\n",
    "  \n",
    "  return tree, joinOrder\n",
    "\n",
    "print(\"I may have done too much work for this problem...not sure\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d2db156c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here is the tree:\n",
      "((0, (4, (5, 6))), (2, (1, 3)))\n",
      "\n",
      "I DID IT!!!!\n"
     ]
    }
   ],
   "source": [
    "# PUT IT ALL TOGETHER\n",
    "\n",
    "from math import exp\n",
    "\n",
    "seqData = ['AGTACGTATCGATGGACCTTA',\n",
    "           'ATCCGTATGCCCTTAATTAGA',\n",
    "           'ATGCGTGCAATGCTATTCCTA',\n",
    "           'TTGCTATTAATGCCCTCCAAA',\n",
    "           'ATCGTAGCTAGACTTTAGTAA',\n",
    "           'ATCGTAGTCTGATTTTAACCA', \n",
    "           'ATCGTAGTCTGATCCCAATTA']\n",
    "\n",
    "DNA = {'G': { 'G': 8, 'C':-5, 'A':-5, 'T':-5, 'N':0 },\n",
    "       'C': { 'G':-5, 'C': 8, 'A':-5, 'T':-5, 'N':0 },\n",
    "       'A': { 'G':-5, 'C':-5, 'A': 8, 'T':-5, 'N':0 },\n",
    "       'T': { 'G':-5, 'C':-5, 'A':-5, 'T': 8, 'N':0 },\n",
    "       'N': { 'G': 0, 'C': 0, 'A': 0, 'T': 0, 'N':0 }} \n",
    "\n",
    "def seqAlign(seqA, seqB, simMatrix=DNA, insert=8, extend=4): \n",
    " \n",
    "\n",
    "  \"\"\"Variables numI and numJ represent the length of the rows and columns of the comparison grid â†’ \n",
    "     needs to be one length longer than each sequence\"\"\"\n",
    "  numI = len(seqA) + 1\n",
    "  numJ = len(seqB) + 1\n",
    "\n",
    "  # Create two matrices. One for holding the scores and one for holding the route taken\n",
    "  scoreMatrix = [[0] * numJ for x in range(numI)]\n",
    "  routeMatrix = [[0] * numJ for x in range(numI)]\n",
    "  \n",
    "  # Make leftmost column have score start with 0 and decrease by 3 for each iteration\n",
    "  for i in range(1, numI):\n",
    "    routeMatrix[i][0] = 1\n",
    "   \n",
    "    \n",
    "  # Make topmost row have score start with 0 and decrease by 3 for each iteration\n",
    "  for j in range(1, numJ):\n",
    "    routeMatrix[0][j] = 2\n",
    "  \n",
    " # Nested loop used to fill in rest of the matrix \n",
    "  for i in range(1, numI):\n",
    "    for j in range(1, numJ):\n",
    "      \n",
    "      # Penalty of a gap is 3 like our example from class \n",
    "      penalty1 = insert\n",
    "      penalty2 = insert\n",
    "      \n",
    "      #NOT SURE HOW TO FIX THIS THOUGH TO MATCH THE 0,-3, -6â€¦.THIS IS WHAT I CAME UP WITH\n",
    "      # Check to see if previous is a gap\n",
    "      if routeMatrix[i-1][j] == 1:\n",
    "        penalty1 = extend\n",
    "        \n",
    "      elif routeMatrix[i][j-1] == 2:\n",
    "        penalty2 = extend\n",
    "       \n",
    "       # Look up similarity score of bases for seqA and seqB \n",
    "      similarity = simMatrix[seqA[i-1] ][seqB[j-1] ]\n",
    "      \n",
    "      # Make a list of three possible scores (left-up diagonal, up, and left)\n",
    "      paths = [scoreMatrix[i-1][j-1] + similarity, # Route 0\n",
    "               scoreMatrix[i-1][j] - penalty1, # Route 1\n",
    "               scoreMatrix[i][j-1] - penalty2] # Route 2                     \n",
    "      \n",
    "      # The best of the three scores are assigned in â€˜bestâ€™ and the index of best score is assigned to route\n",
    "      best = max(paths)\n",
    "      route = paths.index(best)           \n",
    "\n",
    "      # The scores and indexes for each iteration are saved in the matrix \n",
    "      scoreMatrix[i][j] = best\n",
    "      routeMatrix[i][j] = route\n",
    "   \n",
    "   #  Initialize blank lists to hold the optimal sequence alignments for seqA and seqB    \n",
    "  alignA = []\n",
    "  alignB = []\n",
    "\n",
    "  # Assigning variables to implement into while loop below for working backwards\n",
    "  i = numI-1\n",
    "  j = numJ-1\n",
    "  score = scoreMatrix[i][j]\n",
    "\n",
    "  while i > 0 or j > 0:\n",
    "    route = routeMatrix[i][j]    \n",
    "\n",
    "     # Determining which of the three routes had the highest score\n",
    "    if route == 0: # Diagonal\n",
    "      alignA.append( seqA[i-1] )\n",
    "      alignB.append( seqB[j-1] )\n",
    "      i -= 1\n",
    "      j -= 1\n",
    "\n",
    "    elif route == 1: # Gap in seqB\n",
    "      alignA.append(seqA[i-1] )\n",
    "      alignB.append('-')\n",
    "      i -= 1      \n",
    "\n",
    "    elif route == 2: # Gap in seqA\n",
    "      alignA.append('-')\n",
    "      alignB.append(seqB[j-1] ) \n",
    "      j -= 1\n",
    "  \n",
    "   # Reverse the alignment lists since we originally stored them backwards, now we want them forwards \n",
    "  alignA.reverse()\n",
    "  alignB.reverse()\n",
    "  alignA = ''.join(alignA)\n",
    "  alignB = ''.join(alignB)\n",
    "\n",
    "  return score, alignA, alignB \n",
    "\n",
    "def calcSeqSimilarity(seqA, seqB, simMatrix):\n",
    "\n",
    "  numPlaces = min(len(seqA), len(seqB))\n",
    "  \n",
    "  totalScore = 0.0\n",
    "  \n",
    "  for i in range(numPlaces):\n",
    "    \n",
    "    residueA = seqA[i]\n",
    "    residueB = seqB[i]\n",
    "  \n",
    "    totalScore += simMatrix[residueA][residueB]\n",
    "\n",
    "  return totalScore\n",
    "\n",
    "def getDistanceMatrix(seqs, simMatrix):\n",
    "\n",
    "  numbers = len(seqs) #variable numbers for storing sequence lengths\n",
    "  matrix = [[0.0] * numbers for x in range(numbers)]\n",
    "  maxScores = [calcSeqSimilarity(x, x, simMatrix) for x in seqs]\n",
    "  \n",
    "  # Sequences are aligned using an input subsititution matrix\n",
    "  for i in range(numbers-1):\n",
    "    seqA = seqs[i]\n",
    "  \n",
    "    for j in range(i+1,numbers):\n",
    "      seqB = seqs[j]\n",
    "      \n",
    "      # Alignment is passed back with a score which then gets converted into a distance\n",
    "      score, alignA, alignB = seqAlign(seqA, seqB, simMatrix)\n",
    "      maxScore = max(maxScores[i],maxScores[j])\n",
    "      dist = maxScore - score\n",
    "      \n",
    "      matrix[i][j] = dist\n",
    "      matrix[j][i] = dist\n",
    "\n",
    "  return matrix\n",
    "\n",
    "def getJoinPair(distMatrix):\n",
    "\n",
    "  numbers = len(distMatrix) #variable for holding length of matrix\n",
    "  \n",
    "  minQ = None\n",
    "  joinPair = None\n",
    "\n",
    "  # nested loop for iterating through matrix  \n",
    "  for i in range(numbers-1):\n",
    "    sumRow = sum(distMatrix[i])\n",
    "  \n",
    "    for j in range(i+1, numbers):\n",
    "      sumCol = sum(distMatrix[j])\n",
    "    \n",
    "      dist = distMatrix[i][j]\n",
    "      q = (numbers-2)*dist - sumRow - sumCol # Most important here is how we calculate q\n",
    "      \n",
    "      if (minQ is None) or (q < minQ):\n",
    "        minQ = q\n",
    "        joinPair = [i,j] # If smallest so far, we save it as the joined pair aka closest sequences\n",
    "\n",
    "  return joinPair\n",
    "\n",
    "\n",
    "# Next function is to determine distance from a point on tree to the branch point...helps find distance to nodes\n",
    "def getDistToJunction(distMatrix, i, j):\n",
    "  \n",
    "  numbers = len(distMatrix) #variable for holding length of matrix\n",
    "  row = distMatrix[i] # i is row\n",
    "  column = distMatrix[j] # j is column\n",
    " \n",
    "  dist = distMatrix[i][j] + (sum(row)-sum(column))/(numbers-2) # calculate distance\n",
    "  dist *= 0.5 #take that and divide by 2\n",
    "\n",
    "  return dist\n",
    "\n",
    "# Last function builds the tree\n",
    "def neighborJoinTree(distMatrix):\n",
    " \n",
    "  joinOrder = [] #empty list\n",
    "  numbers = len(distMatrix) #variable for holding length of matrix\n",
    "  tree = list(range(numbers))  # list needed for Python 3\n",
    "  \n",
    "  # Using a while loop to iterate while numbers is greater than 2\n",
    "  while numbers > 2:\n",
    "\n",
    "    x, y = getJoinPair(distMatrix)\n",
    "\n",
    "    # New branch point is joined\n",
    "    node = (tree[x], tree[y])\n",
    "    joinOrder.append(node)\n",
    "    tree.append(node)\n",
    "\n",
    "    del tree[y]\n",
    "    del tree[x]\n",
    "\n",
    "    distX = getDistToJunction(distMatrix, x, y)\n",
    "    distY = getDistToJunction(distMatrix, y, x)\n",
    "  \n",
    "    distMatrix.append([0] * (numbers+1))\n",
    "\n",
    "    for i in range(numbers):\n",
    "      if i not in (x,y):\n",
    "\n",
    "        dist = (distMatrix[x][i]-distX) + (distMatrix[y][i]-distY)\n",
    "        dist *= 0.5\n",
    "  \n",
    "        distMatrix[i].append(dist)\n",
    "        distMatrix[numbers][i] = dist\n",
    "\n",
    "    del distMatrix[y]\n",
    "    del distMatrix[x]\n",
    "  \n",
    "    for row in distMatrix:\n",
    "      del row[y]\n",
    "      del row[x]\n",
    "\n",
    "    numbers -= 1\n",
    "\n",
    "  tree = tuple(tree) #Change tree to a tuple\n",
    "  joinOrder.append(tree)\n",
    "  \n",
    "  return tree, joinOrder\n",
    "\n",
    "#TEST TO SEE IF IT WORKS\n",
    "\n",
    "distMatrix = getDistanceMatrix(seqData, DNA)\n",
    "\n",
    "tree, treeJoiner = neighborJoinTree(distMatrix)\n",
    "\n",
    "print('Here is the tree:')\n",
    "print(tree)\n",
    "print('')\n",
    "print(\"I DID IT!!!!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78215dff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
